### PR TITLE
Make ParserContext generic, with a Parser trait.

### PR DESCRIPTION
**Pull request by SimonSapin**
r? @nox

---
This change is [Reviewable](https://reviewable.io/reviews/servo/rust-selectors/101)

### CODE REVIEW

File: src/parser.rs [hunk 1]
```diff
@@ -137,6 +137,21 @@ impl<Impl: SelectorImpl> ParserContext<Impl> {
 }
 
 #[cfg_attr(feature = "heap_size", derive(HeapSizeOf))]
+#[derive(PartialEq, Clone, Debug)]
+pub struct SelectorList<Impl: SelectorImpl>(pub Vec<Selector<Impl>>);
+
+impl<Impl: SelectorImpl> SelectorList<Impl> {
+    /// Parse a comma-separated list of Selectors.
+    /// aka Selector Group in http://www.w3.org/TR/css3-selectors/#grouping
```
Comment (nox) [inline src/parser.rs]: https://drafts.csswg.org/selectors/#grouping would be better no?

File: src/parser.rs [hunk 2]
```diff
@@ -342,6 +357,20 @@ impl<Impl: SelectorImpl> Debug for LocalName<Impl> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { self.to_css(f) }
 }
 
+impl<Impl: SelectorImpl> ToCss for SelectorList<Impl> {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        let mut iter = self.0.iter();
+        let first = iter.next()
+            .expect("Empty SelectorList, should contain at least one selector");
```
Comment (nox) [inline src/parser.rs:358]: What about returning `Err(fmt::Error)`?

File: src/parser.rs [hunk 3]
```diff
@@ -1179,16 +1196,16 @@ pub mod tests {
...
     fn parse_ns(input: &str, context: &ParserContext<DummySelectorImpl>)
-                -> Result<Vec<Selector<DummySelectorImpl>>, ()> {
-        let result = parse_selector_list(context, &mut CssParser::new(input));
+                -> Result<SelectorList<DummySelectorImpl>, ()> {
+        let result = SelectorList::parse(context, &mut CssParser::new(input));
         if let Ok(ref selectors) = result {
-            assert_eq!(selectors.len(), 1);
-            assert_eq!(selectors[0].to_css_string(), input);
+            assert_eq!(selectors.0.len(), 1);
+            assert_eq!(selectors.0[0].to_css_string(), input);
```
Comment (nox) [inline src/parser.rs:1225]: Could you deref to the Vec instead?

File: src/parser.rs [hunk 4]
```diff
@@ -120,19 +97,35 @@ with_heap_size_bound!(+ ::heapsize::HeapSizeOf);
 #[cfg(not(feature = "heap_size"))]
 with_heap_size_bound!();
 
-pub struct ParserContext<Impl: SelectorImpl> {
-    pub in_user_agent_stylesheet: bool,
-    pub default_namespace: Option<Impl::NamespaceUrl>,
-    pub namespace_prefixes: HashMap<Impl::NamespacePrefix, Impl::NamespaceUrl>,
-}
+pub trait Parser {
+    type Impl: SelectorImpl;
```
Comment (nox) [inline src/parser.rs:101]: Could we merge it with `SelectorImpl`?

File: src/parser.rs [hunk 5]
```diff
@@ -986,26 +986,27 @@ fn parse_compound_selector<P, Impl>(
 
 fn parse_functional_pseudo_class<P, Impl>(parser: &P,
                                           input: &mut CssParser,
-                                          name: &str,
+                                          name: Cow<str>,
```
Comment (nox) [inline src/parser.rs:989]: What's the use for this?

**Review by nox — CHANGES_REQUESTED** [review]
Some questions and nits, r=me otherwise.

File: src/parser.rs (hunk 1 above)
Comment (SimonSapin) [reply src/parser.rs]: Done.

File: src/parser.rs (hunk 5 above)
Comment (SimonSapin) [reply src/parser.rs:989]: In case this callback wants to take ownership of a `String` for the name.

File: src/parser.rs (hunk 2 above)
Comment (SimonSapin) [reply src/parser.rs:358]: I don’t think that would be appropriate, `fmt::Error` is for writing to the stream fails. And a selector list really should not be empty. (But that’s not worth enforcing with the type system.)

File: src/parser.rs (hunk 3 above)
Comment (SimonSapin) [reply src/parser.rs:1225]: It could, but I don’t think it’s appropriate as it’s not a pointer.

**Pull request merged**